# AsmViz

(That's the tentative name for this, at least)

An interactive visualization tool for genomic assembly graphs. The goal
of this, as compared with other visualization tools, is to show the
pertinent parts of a graph instead of just displaying the entire graph at once.

To this end, AsmViz highlights certain patterns of contigs in the graph
(bubbles, frayed ropes, chains, and cycles), splits graphs up by connected
components (optionally displaying only a certain maximum number of
components and/or only displaying components with a certain minimum number
of contigs), and uses [GraphViz](http://www.graphviz.org/) to lay out each
component.

AsmViz is composed of two main components:

1. graph\_collator, a Python script that reads an assembly graph file,
   identifies patterns in it, separates it by connected components, and
   runs GraphViz on each component to be laid out (generating a .xdot
   layout file and, optionally, a .dot intermediate file).
2. xdot2cy, a Javascript program that reads an xdot file generated by
   GraphViz and renders the resulting graph in
   [Cytoscape.js](http://js.cytoscape.org/). This is coupled with an
   interface and "control panel" in which the graph can be searched,
   zoomed, panned, and fitted to the screen, nodes can be selected,
   and pattern-indicating groups of nodes can be collapsed (either on
   the level of individual nodes or for all node groups in the graph).
