# MetagenomeScope

An interactive visualization tool designed for metagenomic sequence assembly
and scaffold graphs. The goal
of this, as compared with other visualization tools, is to show the
pertinent parts of a graph instead of just displaying the entire graph at once.

To this end, MetagenomeScope highlights certain patterns of contigs in the graph
(bubbles, frayed ropes, chains, and "linear" cycles), splits graphs up by
connected components, and uses [Graphviz](http://www.graphviz.org/)' `dot` tool
to hierarchically lay out each connected component of an assembly graph.

MetagenomeScope is composed of two main components:

1. `collate.py`, a Python script that reads an assembly graph file,
   identifies patterns in it, separates it by connected components, and
   runs Graphviz on each component to be laid out, generating a SQLite
   .db file containing layout, pattern, and biological information extracted
   from Graphviz and from the original assembly graph file.
   Currently, this supports LastGraph (Velvet), GML
   ([MetaCarvel](https://github.com/marbl/bambus3)), and GFA input
   files, and support for GFA2
   and FASTG (SPAdes) files is planned. Please note that, although
   GFA files only have DNA sequences as an optional requirement
   (it's possible to denote the DNA of a sequence with just an asterisk),
   including DNA sequences in GFA files is currently required.

2. `viewer.html`, a HTML/Javascript webpage that reads a .db file
   generated by `collate.py` and renders the resulting graph in
   [Cytoscape.js](http://js.cytoscape.org/). This is coupled with an
   interface and "control panel" in which the graph can be searched,
   zoomed, panned, rotated, and fitted to the screen, nodes can be selected,
   and pattern-indicating groups of nodes can be collapsed (either on
   the level of individual nodes or for all node groups in the graph). The
   split nature of `collate.py` and `viewer.html` here allows the
   user greater latitude:
    - The user can save the layout/pattern information in a .db file and
      view it at a later time, without incurring the layout and
      patttern detection costs twice
    - The user can host `viewer.html` and a number of .db files on
      their server, allowing many users to view the assembly graphs without
      of the costs associated with layout and structural pattern detection

## System Requirements

### collate.py

* [Python 2.7](https://www.python.org/)
* [PyGraphviz](https://pygraphviz.github.io/)
* [Graphviz](http://graphviz.org/), with the `dot` and `sfdp` layout programs installed

### viewer.html

* Any modern internet browser (smartphone/tablet mobile browsers should
  work, also) supported by Cytoscape.js
    * Google Chrome and Mozilla Firefox are recommended, but others should
      be fine. If you run into any problems using the viewer application on
      any browser, please let me know and I can look into it.

## Running collate.py

`collate.py` is located in the graph\_collator folder. It can be
run from the command line;
see the [system requirements](#system-requirements) section above
for information on what software needs to be installed.

Running `collate.py` will process an assembly/scaffold graph file so that
it can be visualized. The syntax for this is

`./collate.py [-h] -i INPUTFILE -o OUTPUTPREFIX [-d OUTPUTDIRECTORY] [-pg]
    [-px] [-w] [-b BICOMPONENTSFILE]`

### Script output

The script will always produce a `.db` file that can be loaded in the viewer
application to visualize the assembly graph.

If the `-pg` argument is passed, `.gv` files (in the DOT language)
for each connected component of the assembly graph will be generated; if the
`-px` argument is passed, `.xdot` files (in the xdot language) for each
connected component of the assembly graph will be generated.

The script will also generate a few types of auxiliary files containing various
information about the structure of the assembly graph. These files are:

* `prefix_links`, where `prefix` is the output prefix passed via `-o`. Only one
  of these files will be generated per execution of `collate.py`. This file
  indicates all the edges in the assembly graph. If you pass in `-b` and the
  input assembly graph has unoriented contigs, then this file will not be
  generated (since it would be equivalent to the _single_links file in that
  case).
* `prefix_single_links`, where `prefix` is the output prefix passed via `-o`.
  This file will only be generated if the input assembly graph has unoriented
  contigs. In terms of currently supported input filetypes, this means that
  this file will only be generated when the input assembly graph is of type
  LastGraph or GFA.
* `prefix_bicmps`, where `prefix` is the output prefix passed via `-o`. Only
  one of these files will be generated per execution of `collate.py`. This
  file indicates the various separation pairs contained within the assembly
  graph (see [Nijkamp et al.](https://www.ncbi.nlm.nih.gov/pubmed/24058058)
  for a brief overview of separation pairs and their usage in bubble
  detection). It's possible to pass an existing version of this file using `-b`
  to the script, to prevent having to do the work of creating the file again.
* `component_D.info`, where `D` is an integer greater than 0. There will be one
  of these files created for every biconnected component contained within the
  assembly graph: these files indicate the contents of the SPQR tree defined
  for their corresponding biconnected component.
* `spqrD.gml`, where `D` is an integer greater than 0. These files correspond
  to `component_D.info` files: they indicate the connections between the
  metanodes of a SPQR tree.

The script requires all `component_D.info` and `spqrD.gml` files to be
removed from the output directory before it generates more of them.
If `-w` is enabled, then all existing files with corresponding names in the
output directory will be deleted; however, if `-w` is not enabled, then an
error will be raised.

Similarly, if files exist in the output directory with filenames overlapping 
those of the `prefix_links` and `prefix_bicmps` files, then those files will be
either deleted (if `-w` is enabled) or an error will be raised (if `-w` is not
enabled).

### Command-line argument descriptions

* `-i` The input assembly graph file to be used.
* `-o` The file prefix to be used for all files generated. As an example, given
  the argument
  `-o prefix`, the file `prefix.db` would be generated. If .gv and/or .xdot
  files are created (depending on the `-pg` or `-px` arguments, respectively),
  then those files will be numbered according to the relative size rank
  (in nodes) of their respective connected component within the assembly graph.
* `-d` This optional argument specifies the name of the directory in which
  all output files will be stored. If this argument is not indicated, then all
  files will be generated in the current working directory.
* `-pg` This optional argument produces DOT files (suffix .gv) in the output
  directory. As an example, given the arguments `-o prefix` and `-pg` for an
  assembly graph with 3 connected components, the files `prefix.db`,
  `prefix_1.gv`, `prefix_2.gv`, and `prefix_3.gv` would be created (where
  `prefix_1.gv` indicates the largest connected component by number of nodes,
  `prefix_2.gv` indicates the next largest connected component, and so on).
* `-px` This optional argument produces .xdot files in the output
  directory. These files are labelled in an identical fashion to `.gv` files,
  with the only difference in naming being the file suffix (.xdot instead of
  .gv).
* `-b` This optional argument lets you pass in an existing file indicating the
  separation pairs in the graph (to be used in the detection of complex
  bubbles) to the script.
* `-w` This optional argument allows the overwriting of output files
  (.db/.xdot/.gv/links/single_links/bicmps/.info/spqr.gml files).
  If this argument is **not** given, then:
    * An error will be raised if writing a .db file would cause another
      .db file to be overwritten.
    * A warning will be displayed if writing to a .gv or .xdot file would cause
      another .gv/.xdot file to be overwritten. In this case, the .gv/.xdot
      file in question simply would not be saved.
    * Note that the presence of files in the
      output directory that are conflicting-named folders (e.g. a
      directory named `e_coli.db/` in the output directory while attempting to
      produce a file named `e_coli.db`) will cause an error/warning to be
      raised regardless of whether or not `-w` is set.

## Running viewer.html

TODO

## Acknowledgements: tools used

* `collate.py` uses [Graphviz](http://www.graphviz.org/)' dot and sfdp layout programs via the
  [PyGraphviz](http://pygraphviz.github.io/) interface.
* Both `collate.py` and the MetagenomeScope viewer use [sqlite3](https://sqlite.org/).
  In particular, `collate.py` uses the aforementioned Python sqlite3 module
  while the MetagenomeScope viewer uses [sql.js](https://github.com/kripken/sql.js/) to
  process .db files on the client side.
* The MetagenomeScope viewer uses [Cytoscape.js](https://js.cytoscape.org/) to render
  graphs on the client side.
    * Also, the toggling protocol used for the control panel of the MetagenomeScope
      viewer was inspired by a similar toggling mechanism used in
      [this Cytoscape.js
      demo](http://js.cytoscape.org/demos/2ebdc40f1c2540de6cf0/).
* The MetagenomeScope viewer uses [jQuery](https://jquery.com/) and
  [Bootstrap](http://getbootstrap.com/) for various stylistic and functional
  purposes in the application.
    * The icons used to theme various controls in the viewer application are
      from the [Glyphicon](http://glyphicons.com/) Halflings set,
      included with Bootstrap.
